src/domain/address.dataloader.ts

```diff
@@ -0,0 +1,25 @@
+import * as DataLoader from 'dataloader';
+import { AddressDbDataSource } from '@data/sources/address.db.datasource';
+import { AddressResponse } from '@api/user/address.type';
+import { Service } from 'typedi';
+
+@Service()
+export class AddressLoader {
+  constructor(private readonly datasource: AddressDbDataSource) {}
+
+  exec() {
+    return new DataLoader<string, AddressResponse[] | undefined>(async (userIds) => {
+      const addresses = await this.datasource.findByUserIds(userIds as string[]);
+      const userIdToAddress: Record<string, AddressResponse[]> = {};
+      addresses.forEach((u) => {
+        if (!userIdToAddress[u.userId!]) {
+          userIdToAddress[u.userId!] = [u];
+        } else {
+          userIdToAddress[u.userId!].push(u);
+        }
+      });
+
+      return userIds.map((userId) => userIdToAddress[userId]);
```

Comment:
isso é para pegar o array com os endereços ne? um filter no addresses n resolve?

```suggestion
      return userIds.map((userId) => addresses.filter((address) => address.userId === userId));
```

Reply:
sim, mas fica mais lento.  usando hashTable fica O(N), esse ai fica O(Nˆ2) ou O(MN), considerando que o tamanho do array de users é diferente do de endereços.